Replace `RecursiveMutex cs_mapLocalHost` with Mutex, and rename it

w0xlt opened the pull request:
This PR is related to #19303 and gets rid of the `RecursiveMutex cs_mapLocalHost`.

hebasto commented on src/net.cpp:
```diff
@@ -112,9 +112,9 @@ static const uint64_t RANDOMIZER_ID_ADDRCACHE = 0x1cf2e4ddd306dda9ULL; // SHA256
-RecursiveMutex cs_mapLocalHost;
-std::map<CNetAddr, LocalServiceInfo> mapLocalHost GUARDED_BY(cs_mapLocalHost);
-static bool vfLimited[NET_MAX] GUARDED_BY(cs_mapLocalHost) = {};
+Mutex map_localhost_mutex;
```
Suggesting to add `g_` prefix to explicitly note that this mutex belongs to the global namespace:
```suggestion
Mutex g_maplocalhost_mutex;
```

Also see #19180.

w0xlt replied:
Thanks for suggestion. Done.

hebasto reviewed (commented):
Approach ACK 3e05ab9537dd4478c66b26f295b63e8081a24bc5.

shaavan reviewed (commented):
Approach ACK 3e05ab9537dd4478c66b26f295b63e8081a24bc5

Going commit wise:
1. The renaming of the mutex `cs_mapLocalHost` -> `map_localhost_mutex`, is an improvement, in my opinion.
2. The type of `map_localhost_mutex` is changed from **RecursiveMutex** to **Mutex**. I checked through the codebase where `map_localhost_mutex` is used for any instances of recursive lockings, and I found none. Also, I was able to compile this PR without any UB warnings successfully. So I think it’s safe to convert the type of this mutex from **RecursiveMutex** to **Mutex**.

I shall ACK this PR as soon as [this](https://github.com/bitcoin/bitcoin/pull/24099#discussion_r787419656) suggestion by @hebasto is addressed.

w0xlt committed:
scripted-diff: rename cs_mapLocalHost -> g_maplocalhost_mutex

-BEGIN VERIFY SCRIPT-
s() { sed -i 's/cs_mapLocalHost/g_maplocalhost_mutex/g' $1; }
s src/net.cpp
s src/net.h
s src/rpc/net.cpp
s src/test/net_tests.cpp
-END VERIFY SCRIPT-

w0xlt committed:
refactor: replace RecursiveMutex g_maplocalhost_mutex with Mutex

shaavan reviewed (approved):
ACK 5e7e4c9f6e57f5333bd17a20b0c85a78d032998e

Changes since my last review:

- `map_localhost_mutex` -> `g_maplocalhost_mutex`.

The updated PR compiled successfully on Ubuntu 20.04.

theStack reviewed (approved):
ACK 5e7e4c9f6e57f5333bd17a20b0c85a78d032998e

Checked that within all critical sections that are protected by `g_maplocalhost_mutex`, there is no chance another one is called, i.e. a regular Mutex is sufficient.

hebasto reviewed (approved):
ACK 5e7e4c9f6e57f5333bd17a20b0c85a78d032998e, I have reviewed the code and it looks OK, I agree it can be merged.

In particular, I've verified that there are no recursive mutex locking on the path `GetLocal()` --> `CNetAddr::GetReachabilityFrom()` --> `CNetAddr::IsRoutable()`.

---

A script in the "scripted-diff" commit could be more robust:
```
sed -i 's/cs_mapLocalHost/g_maplocalhost_mutex/g' -- $(git grep --files-with-matches 'cs_mapLocalHost')
```
